File: macos/TeachGateVPN/PacketTunnelProvider.m
// Copyright 2018 The Outline Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "PacketTunnelProvider.h"
#if __has_include("TeachGateVPN-Swift.h")
#import "TeachGateVPN-Swift.h"
#else
// Graceful fallback for indexers or partial builds where the generated Swift header
// is not available yet. Methods are typed as 'id' to avoid import-order issues.
@interface SwiftBridge : NSObject
+ (id)getTunnelNetworkSettings;
+ (id)newInvalidConfigOutlineErrorWithMessage:(NSString *)message;
+ (id)newInternalOutlineErrorWithMessage:(NSString *)message;
+ (id)newOutlineErrorFromNsError:(NSError *)error;
+ (void)saveLastErrorWithNsError:(NSError *)err;
+ (NSData *)loadLastErrorToIPCResponse;
@end
#endif

#include <arpa/inet.h>
#include <ifaddrs.h>
#include <netdb.h>

#import <Foundation/Foundation.h>
#import <NetworkExtension/NetworkExtension.h>
#import <Tun2socks/Tun2socks.h>

#if __has_include(<CocoaLumberjack/CocoaLumberjack.h>)
  #import <CocoaLumberjack/CocoaLumberjack.h>
  #if DEBUG
    static DDLogLevel ddLogLevel = DDLogLevelDebug;
  #else
    static DDLogLevel ddLogLevel = DDLogLevelInfo;
  #endif
  #define TG_DD_AVAILABLE 1
#else
  #define TG_DD_AVAILABLE 0
  // Fallback macros to NSLog when CocoaLumberjack isn't available (e.g., pods not installed yet).
  #ifndef DDLogInfo
    #define DDLogInfo(...) NSLog(__VA_ARGS__)
  #endif
  #ifndef DDLogDebug
    #define DDLogDebug(...) NSLog(__VA_ARGS__)
  #endif
  #ifndef DDLogError
    #define DDLogError(...) NSLog(__VA_ARGS__)
  #endif
  #ifndef DDLogWarn
    #define DDLogWarn(...) NSLog(__VA_ARGS__)
  #endif
#endif

// Not exposed by gobind headers in this build for Objective-C.
// Declare as a weak-import symbol so the binary links even if the symbol is absent.
FOUNDATION_EXTERN __attribute__((weak_import)) Tun2socksConnectOutlineTunnelResult*
Tun2socksConnectOutlineTunnel(id<Tun2socksTunWriter> writer, id client, BOOL isUdpSupported);

NSString *const kDefaultPathKey = @"defaultPath";

@interface PacketTunnelProvider ()<Tun2socksTunWriter>
@property (nonatomic, strong, nullable) id<Tun2socksTunnel> tunnel;
@property (nonatomic, copy, nullable) void (^startCompletion)(NSNumber *);
@property (nonatomic, copy, nullable) void (^stopCompletion)(NSNumber *);
@property (nonatomic, copy, nullable) NSString *tunnelId;
@property (nonatomic, copy, nullable) NSString *transportConfig;
@property (nonatomic, strong) dispatch_queue_t packetQueue;
@property (nonatomic, assign) BOOL isUdpSupported;
@end

@implementation PacketTunnelProvider

- (id)init {
  self = [super init];
  if (self) {
    [self setupLogging];
    // Initialize a packet queue for reading/writing packets.
    _packetQueue = dispatch_queue_create("com.teachgatedesk.develentcorp.packetqueue", DISPATCH_QUEUE_SERIAL);
  }
  return self;
}

- (void)setupLogging {
  @try {
#if TG_DD_AVAILABLE
    NSFileManager *fm = [NSFileManager defaultManager];
    NSURL *containerURL = [fm containerURLForSecurityApplicationGroupIdentifier:@"group.com.teachgatedesk.develentcorp"];
    if (!containerURL) {
      // Fallback: still enable OS logger so at least we have syslog
      [DDLog addLogger:[DDOSLogger sharedInstance]];
      DDLogWarn(@"App Group container URL is nil; file logging disabled");
      return;
    }
    NSString *logsPath = [[containerURL.path stringByAppendingPathComponent:@"Logs/Extension"] stringByStandardizingPath];
    [fm createDirectoryAtPath:logsPath withIntermediateDirectories:YES attributes:nil error:nil];

    DDLogFileManagerDefault *logFileManager = [[DDLogFileManagerDefault alloc] initWithLogsDirectory:logsPath];
    DDFileLogger *fileLogger = [[DDFileLogger alloc] initWithLogFileManager:logFileManager];
    fileLogger.rollingFrequency = 60 * 60 * 24; // 24 hours
    fileLogger.maximumFileSize = 5 * 1024 * 1024; // 5 MB
    fileLogger.logFileManager.maximumNumberOfLogFiles = 7;
    fileLogger.logFileManager.logFilesDiskQuota = 30 * 1024 * 1024; // 30 MB

    [DDLog addLogger:[DDOSLogger sharedInstance]];
    [DDLog addLogger:fileLogger];

    DDLogInfo(@"CocoaLumberjack initialized for extension. Logs at %@", logsPath);
#else
    NSLog(@"CocoaLumberjack not available; using NSLog only for extension logging");
#endif
  } @catch (NSException *exception) {
    NSLog(@"Failed to setup CocoaLumberjack logging: %@", exception);
  }
}

- (void)startTunnelWithOptions:(NSDictionary *)options
            completionHandler:(void (^)(NSError *))completion {
  DDLogInfo(@"Starting tunnel");
  DDLogDebug(@"Options are %@", options);

  // mimics fetchLastDisconnectErrorWithCompletionHandler on older systems
  void (^startDone)(NSError *) = ^(NSError *err) {
    [SwiftBridge saveLastErrorWithNsError:err];
    completion(err);
  };

  // MARK: Process Config.
  if (self.protocolConfiguration == nil) {
    DDLogError(@"Failed to retrieve NETunnelProviderProtocol.");
    return startDone([SwiftBridge newInvalidConfigOutlineErrorWithMessage:@"no config specified"]);
  }
  NETunnelProviderProtocol *protocol = (NETunnelProviderProtocol *)self.protocolConfiguration;
  NSString *tunnelId = protocol.providerConfiguration[@"id"];
  if (![tunnelId isKindOfClass:[NSString class]]) {
    DDLogError(@"Failed to retrieve the tunnel id.");
    return startDone([SwiftBridge newInternalOutlineErrorWithMessage:@"no tunnel ID specified"]);
  }

  NSString *transportConfig = protocol.providerConfiguration[@"transport"];
  if (![transportConfig isKindOfClass:[NSString class]]) {
    DDLogError(@"Failed to retrieve the transport configuration.");
    return startDone([SwiftBridge newInvalidConfigOutlineErrorWithMessage:@"config is not a String"]);
  }
  self.tunnelId = tunnelId;
  self.transportConfig = transportConfig;

  // startTunnel has 3 cases:
  // - When started from the app, we get options != nil, with no ["is-on-demand"] entry.
  // - When started on-demand, we get option != nil, with ["is-on-demand"] = 1;.
  // - When started from the VPN settings, we get options == nil
  NSNumber *isOnDemandNumber = options == nil ? nil : options[@"is-on-demand"];
  bool isOnDemand = isOnDemandNumber != nil && [isOnDemandNumber intValue] == 1;
  DDLogDebug(@"isOnDemand is %d", isOnDemand);

  // Adaptation note:
  // The production Outline code performs TCP/UDP connectivity checks using additional Go-bound APIs
  // (OutlineClientConfig, OutlineCheckTCPAndUDPConnectivity, PlatformError, etc.). These are not
  // present in this project build. To integrate with the migrated tun2socks for real VPN traffic
  // while keeping scope to the PacketTunnelProvider, we bypass those pre-checks and proceed to
  // configure the system routes and start tun2socks directly.
  self.isUdpSupported = YES;

  [self startRouting:[SwiftBridge getTunnelNetworkSettings]
          completion:^(NSError *_Nullable error) {
            if (error != nil) {
              return startDone([SwiftBridge newOutlineErrorFromNsError:error]);
            }
            NSError *tun2socksError = [self startTun2Socks:self.isUdpSupported];
            if (tun2socksError != nil) {
              return startDone(tun2socksError);
            }
            [self listenForNetworkChanges];
            startDone(nil);
          }];
}

- (void)stopTunnelWithReason:(NEProviderStopReason)reason
          completionHandler:(void (^)(void))completionHandler {
  DDLogInfo(@"Stopping tunnel, reason: %ld", (long)reason);
  [self stopListeningForNetworkChanges];
  [self.tunnel disconnect];
  [self cancelTunnelWithError:nil];
  completionHandler();
}

# pragma mark - Network

- (void)startRouting:(NEPacketTunnelNetworkSettings *)settings
           completion:(void (^)(NSError *))completionHandler {
  PacketTunnelProvider * __unsafe_unretained weakSelf = self;
  [self setTunnelNetworkSettings:settings completionHandler:^(NSError * _Nullable error) {
    if (error != nil) {
      DDLogError(@"Failed to start routing: %@", error.localizedDescription);
    } else {
      DDLogInfo(@"Routing started");
      // Passing nil settings clears the tunnel network configuration. Indicate to the system that
      // the tunnel is being re-established if this is the case.
      weakSelf.reasserting = settings == nil;
    }
    completionHandler(error);
  }];
}

// Registers KVO for the `defaultPath` property to receive network connectivity changes.
- (void)listenForNetworkChanges {
  [self stopListeningForNetworkChanges];
  [self addObserver:self
         forKeyPath:kDefaultPathKey
            options:NSKeyValueObservingOptionOld
            context:nil];
}

// Unregisters KVO for `defaultPath`.
- (void)stopListeningForNetworkChanges {
  @try {
    [self removeObserver:self forKeyPath:kDefaultPathKey];
  } @catch (id exception) {
    // Observer not registered, ignore.
  }
}

- (void)observeValueForKeyPath:(nullable NSString *)keyPath
                      ofObject:(nullable id)object
                        change:(nullable NSDictionary<NSString *, id> *)change
                       context:(nullable void *)context {
  if (![kDefaultPathKey isEqualToString:keyPath]) {
    return;
  }
  // Guard against false positives by comparing the paths' string description, which includes
  // properties not exposed by the class.
  NWPath *lastPath = change[NSKeyValueChangeOldKey];
  if (lastPath == nil || [lastPath isEqualToPath:self.defaultPath] ||
      [lastPath.description isEqualToString:self.defaultPath.description]) {
    return;
  }

  dispatch_async(dispatch_get_main_queue(), ^{
    [self handleNetworkChange:self.defaultPath];
  });
}

- (void)handleNetworkChange:(NWPath *)newDefaultPath {
  DDLogInfo(@"Network connectivity changed");
  if (newDefaultPath.status == NWPathStatusSatisfied) {
    DDLogInfo(@"Reconnecting tunnel.");
    // Check whether UDP support has changed with the network, if supported by the tunnel object.
    BOOL isUdpSupported = self.tunnel ? [self.tunnel updateUDPSupport] : self.isUdpSupported;
    DDLogDebug(@"UDP support: %d -> %d", self.isUdpSupported, isUdpSupported);
    self.isUdpSupported = isUdpSupported;
    [self reconnectTunnel];
  } else {
    DDLogInfo(@"Clearing tunnel settings.");
    [self startRouting:nil completion:^(NSError * _Nullable error) {
      if (error != nil) {
        DDLogError(@"Failed to clear tunnel network settings: %@", error.localizedDescription);
      } else {
        DDLogInfo(@"Tunnel settings cleared");
      }
    }];
  }
}

/**
 Converts a struct sockaddr address |sa| to a string. Expects |maxbytes| to be allocated for |s|.
 @return whether the operation succeeded.
 */
bool getIpAddressString(const struct sockaddr *sa, char *s, socklen_t maxbytes) {
  if (!sa || !s) {
    DDLogError(@"Failed to get IP address string: invalid argument");
    return false;
  }
  switch (sa->sa_family) {
    case AF_INET:
      inet_ntop(AF_INET, &(((struct sockaddr_in *)sa)->sin_addr), s, maxbytes);
      break;
    case AF_INET6:
      inet_ntop(AF_INET6, &(((struct sockaddr_in6 *)sa)->sin6_addr), s, maxbytes);
      break;
    default:
      DDLogError(@"Cannot get IP address string: unknown address family");
      return false;
  }
  return true;
}

#pragma mark - tun2socks

/** Reconfigures routing and the UDP handler on connectivity change. */
- (void)reconnectTunnel {
  if (!self.transportConfig) {
    DDLogError(@"Failed to reconnect tunnel, missing tunnel configuration.");
    return;
  }
  [self startRouting:[SwiftBridge getTunnelNetworkSettings]
         completion:^(NSError *_Nullable error) {
           if (error != nil) {
             [self cancelTunnelWithError:error];
           }
         }];
}

- (BOOL)close:(NSError *_Nullable *)error {
  return YES;
}

- (BOOL)write:(NSData *_Nullable)packet n:(long *)n error:(NSError *_Nullable *)error {
  [self.packetFlow writePackets:@[ packet ] withProtocols:@[ @(AF_INET) ]];
  return YES;
}

// Writes packets from the VPN to the tunnel.
- (void)processPackets {
  typeof(self) __unsafe_unretained weakSelf = self;
  __block long bytesWritten = 0;
  [weakSelf.packetFlow readPacketsWithCompletionHandler:^(NSArray<NSData *> *_Nonnull packets,
                                                          NSArray<NSNumber *> *_Nonnull protocols) {
    for (NSData *packet in packets) {
      if (weakSelf.tunnel) {
        [weakSelf.tunnel write:packet ret0_:&bytesWritten error:nil];
      }
    }
    dispatch_async(weakSelf.packetQueue, ^{
      [weakSelf processPackets];
    });
  }];
}

// Starts or restarts tun2socks with the current configuration.
- (NSError*)startTun2Socks:(BOOL)isUdpSupported {
  BOOL isRestart = self.tunnel != nil && [self.tunnel isConnected];
  if (isRestart) {
    [self.tunnel disconnect];
  }
  PacketTunnelProvider * __unsafe_unretained weakSelf = self;

  // In the production Outline build, a Go Outline client object is created and passed here.
  // That API is not available in this project. Attempt to call the Go-bound symbol if available.
  if (Tun2socksConnectOutlineTunnel == NULL) {
    DDLogError(@"Tun2socksConnectOutlineTunnel symbol not found in Tun2socks.framework");
    return [SwiftBridge newInternalOutlineErrorWithMessage:@"Tun2socksConnectOutlineTunnel not available"];
  }
  Tun2socksConnectOutlineTunnelResult *result =
      Tun2socksConnectOutlineTunnel(weakSelf, nil, isUdpSupported);

  // Some gobind builds omit the 'error' property from the result. We conservatively check tunnel only.
  id<Tun2socksTunnel> tunnel = result ? result.tunnel : nil;
  if (!tunnel) {
    DDLogError(@"Failed to start tun2socks: no tunnel returned");
    return [SwiftBridge newInternalOutlineErrorWithMessage:@"tun2socks failed to return a tunnel"];
  }
  self.tunnel = tunnel;
  if (!isRestart) {
    dispatch_async(self.packetQueue, ^{
      [weakSelf processPackets];
    });
  }
  return nil;
}

#pragma mark - fetch last disconnect error

// TODO: Remove this code once we only support newer systems (macOS 13.0+, iOS 16.0+)
NSString *const kFetchLastErrorIPCName = @"fetchLastDisconnectDetailedJsonError";

- (void)handleAppMessage:(NSData *)messageData completionHandler:(void (^)(NSData * _Nullable))completion {
  // mimics fetchLastDisconnectErrorWithCompletionHandler on older systems
  NSString *ipcName = [[NSString alloc] initWithData:messageData encoding:NSUTF8StringEncoding];
  if (![ipcName isEqualToString:kFetchLastErrorIPCName]) {
    DDLogWarn(@"Invalid Extension IPC call: %@", ipcName);
    return completion(nil);
  }
  completion([SwiftBridge loadLastErrorToIPCResponse]);
}

- (void)cancelTunnelWithError:(nullable NSError *)error {
  [SwiftBridge saveLastErrorWithNsError:error];
  [super cancelTunnelWithError:error];
}

@end

File: macos/TeachGateVPN/PacketTunnelProvider.swift
// Copyright 2024 The Outline Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Use NSLog for VPN extension logging
@inline(__always) func DDLogInfo(_ msg: @autoclosure () -> String) { NSLog("[INFO] %@", msg()) }
@inline(__always) func DDLogDebug(_ msg: @autoclosure () -> String) { NSLog("[DEBUG] %@", msg()) }
@inline(__always) func DDLogWarn(_ msg: @autoclosure () -> String) { NSLog("[WARN] %@", msg()) }
@inline(__always) func DDLogError(_ msg: @autoclosure () -> String) { NSLog("[ERROR] %@", msg()) }
import NetworkExtension

// Define minimal OutlineError stub since the module import is not working
public enum OutlineError: Error {
  case internalError(message: String)
  case invalidConfig(message: String)
  
  public var code: String {
    switch self {
    case .internalError: return "internalError"
    case .invalidConfig: return "invalidConfig"
    }
  }
}

public func toOutlineError(error: Error) -> OutlineError {
  if let outlineError = error as? OutlineError {
    return outlineError
  }
  return OutlineError.internalError(message: error.localizedDescription)
}

public func marshalErrorJson(error: Error) -> String {
  let outlineErr = toOutlineError(error: error)
  return "{\"code\":\"\(outlineErr.code)\",\"message\":\"\(error.localizedDescription)\"}"
}

/// SwiftBridge is a transitional class to allow the incremental migration of our PacketTunnelProvider from Objective-C to Swift.
@objcMembers
public class SwiftBridge: NSObject {

  /** Helper function that we can call from Objective-C. */
  public static func getTunnelNetworkSettings() -> NEPacketTunnelNetworkSettings {
    // The remote address is not required, but needs to be valid, or else you get a
    // "Invalid NETunnelNetworkSettings tunnelRemoteAddress" error.
    let settings = NEPacketTunnelNetworkSettings(tunnelRemoteAddress: "::")

    // Configure VPN address and routing.
    let vpnAddress = selectVpnAddress(interfaceAddresses: getNetworkInterfaceAddresses())
    let ipv4Settings = NEIPv4Settings(addresses: [vpnAddress], subnetMasks: ["255.255.255.0"])
    ipv4Settings.includedRoutes = [NEIPv4Route.default()]
    ipv4Settings.excludedRoutes = getExcludedIpv4Routes()
    settings.ipv4Settings = ipv4Settings

    // Configure with Cloudflare, Quad9, and OpenDNS resolver addresses.
    settings.dnsSettings = NEDNSSettings(servers: [
      "1.1.1.1", "9.9.9.9", "208.67.222.222", "208.67.220.220",
    ])

    return settings
  }

  /**
   Creates a NSError (of `OutlineError.errorDomain`) from the `OutlineError.internalError`.
   */
  public static func newInternalOutlineError(message: String) -> NSError {
    return OutlineError.internalError(message: message) as NSError
  }

  /**
   Creates a NSError (of `OutlineError.errorDomain`) from the `OutlineError.invalidConfig` error.
   */
  public static func newInvalidConfigOutlineError(message: String) -> NSError {
    return OutlineError.invalidConfig(message: message) as NSError
  }

  /**
   Creates a NSError (of `OutlineError.errorDomain`) with detailed JSON from another NSError.
   */
  public static func newOutlineErrorFrom(nsError: Error?) -> NSError? {
    guard let nserr = nsError else {
      return nil
    }
    return toOutlineError(error: nserr) as NSError
  }

  // TODO: Remove this code once we only support newer systems (macOS 13.0+, iOS 16.0+)
  public static func saveLastError(nsError: Error?) {
    saveLastDisconnectErrorDetails(error: nsError)
  }

  // TODO: Remove this code once we only support newer systems (macOS 13.0+, iOS 16.0+)
  public static func loadLastErrorToIPCResponse() -> NSData? {
    return loadLastDisconnectErrorDetailsToIPCResponse() as? NSData
  }
}

// Represents an IP subnetwork.
// Note that this class and its non-private properties must be public in order to be visible to the ObjC
// target of the OutlineAppleLib Swift Package.
class Subnet: NSObject {
  // Parses a CIDR subnet into a Subnet object. Returns nil on failure.
  public static func parse(_ cidrSubnet: String) -> Subnet? {
    let components = cidrSubnet.components(separatedBy: "/")
    guard components.count == 2 else {
      NSLog("Malformed CIDR subnet")
      return nil
    }
    guard let prefix = UInt16(components[1]) else {
      NSLog("Invalid subnet prefix")
      return nil
    }
    return Subnet(address: components[0], prefix: prefix)
  }

  public var address: String
  public var prefix: UInt16
  public var mask: String

  public init(address: String, prefix: UInt16) {
    self.address = address
    self.prefix = prefix
    let mask = (0xffff_ffff as UInt32) << (32 - prefix)
    self.mask = mask.IPv4String()
  }
}

extension UInt32 {
  // Returns string representation of the integer as an IP address.
  public func IPv4String() -> String {
    let ip = self
    let a = UInt8((ip >> 24) & 0xff)
    let b = UInt8((ip >> 16) & 0xff)
    let c = UInt8((ip >> 8) & 0xff)
    let d = UInt8(ip & 0xff)
    return "\(a).\(b).\(c).\(d)"
  }
}

// Returns all IPv4 addresses of all interfaces.
func getNetworkInterfaceAddresses() -> [String] {
  var interfaces: UnsafeMutablePointer<ifaddrs>?
  var addresses = [String]()

  guard getifaddrs(&interfaces) == 0 else {
    DDLogError("Failed to retrieve network interface addresses")
    return addresses
  }

  var interface = interfaces
  while interface != nil {
    // Only consider IPv4 interfaces.
    if interface!.pointee.ifa_addr.pointee.sa_family == UInt8(AF_INET) {
      let addr = interface!.pointee.ifa_addr!.withMemoryRebound(to: sockaddr_in.self, capacity: 1) {
        $0.pointee.sin_addr
      }
      if let ip = String(cString: inet_ntoa(addr), encoding: .utf8) {
        addresses.append(ip)
      }
    }
    interface = interface!.pointee.ifa_next
  }

  freeifaddrs(interfaces)

  return addresses
}

let kVpnSubnetCandidates: [String: String] = [
  "10": "10.111.222.0",
  "172": "172.16.9.1",
  "192": "192.168.20.1",
  "169": "169.254.19.0",
]

// Given the list of known interface addresses, returns a local network IP address to use for the VPN.
func selectVpnAddress(interfaceAddresses: [String]) -> String {
  var candidates = kVpnSubnetCandidates

  for address in interfaceAddresses {
    for subnetPrefix in kVpnSubnetCandidates.keys {
      if address.hasPrefix(subnetPrefix) {
        // The subnet (not necessarily the address) is in use, remove it from our list.
        candidates.removeValue(forKey: subnetPrefix)
      }
    }
  }
  guard !candidates.isEmpty else {
    // Even though there is an interface bound to the subnet candidates, the collision probability
    // with an actual address is low.
    return kVpnSubnetCandidates.randomElement()!.value
  }
  // Select a random subnet from the remaining candidates.
  return candidates.randomElement()!.value
}

let kExcludedSubnets = [
  "10.0.0.0/8",
  "100.64.0.0/10",
  "169.254.0.0/16",
  "172.16.0.0/12",
  "192.0.0.0/24",
  "192.0.2.0/24",
  "192.31.196.0/24",
  "192.52.193.0/24",
  "192.88.99.0/24",
  "192.168.0.0/16",
  "192.175.48.0/24",
  "198.18.0.0/15",
  "198.51.100.0/24",
  "203.0.113.0/24",
  "240.0.0.0/4",
]

func getExcludedIpv4Routes() -> [NEIPv4Route] {
  var excludedIpv4Routes = [NEIPv4Route]()
  for cidrSubnet in kExcludedSubnets {
    if let subnet = Subnet.parse(cidrSubnet) {
      let route = NEIPv4Route(destinationAddress: subnet.address, subnetMask: subnet.mask)
      excludedIpv4Routes.append(route)
    }
  }
  return excludedIpv4Routes
}

// MARK: - fetch last disconnect error

// TODO: Remove this code once we only support newer systems (macOS 13.0+, iOS 16.0+)

/**
  In the app, we need to use [NEVPNConnection fetchLastDisconnectErrorWithCompletionHandler] to
  retrive the most recent error that caused the VPN extension to disconnect.
  But it's only available on newer systems (macOS 13.0+, iOS 16.0+), so we need a workaround for
  older ones.
  The workaround lets the app to use [NETunnelProviderSession sendProviderMessage] to get the
  error through an IPC method.
  The extension also needs to save the last error to disk, as the system will unload the extension
  after a failed connection.
  We use [NSUserDefaults standardUserDefaults] to store the error, so it's available even after
  the extension restarts.
*/

let lastDisconnectErrorPersistenceKey = "lastDisconnectError"

/// Keep it in sync with the data type defined in OutlineVpn.Swift
/// Also keep in mind that we will always use PropertyListEncoder and PropertyListDecoder to marshal this data.
private struct LastErrorIPCData: Codable {
  let errorCode: String
  let errorJson: String
}

func saveLastDisconnectErrorDetails(error: Error?) {
  guard let err = error else {
    return UserDefaults.standard.removeObject(forKey: lastDisconnectErrorPersistenceKey)
  }
  let outlineErr = toOutlineError(error: err)
  let persistObj = LastErrorIPCData(
    errorCode: outlineErr.code, errorJson: marshalErrorJson(error: err))
  do {
    let encodedObj = try PropertyListEncoder().encode(persistObj)
    UserDefaults.standard.setValue(encodedObj, forKey: lastDisconnectErrorPersistenceKey)
  } catch {
    DDLogError("failed to persist lastDisconnectError \(persistObj): \(error)")
  }
}

func loadLastDisconnectErrorDetailsToIPCResponse() -> Data? {
  return UserDefaults.standard.data(forKey: lastDisconnectErrorPersistenceKey)
}

// MARK: - NEPacketTunnelProvider implementation

private enum ExtensionIPCCommand {
  static let fetchLastDetailedJsonError = "fetchLastDisconnectDetailedJsonError"
}

// Darwin AF_ constants for NEPacketTunnelFlow writeProtocols
private let AF_INET_NUM: Int32 = 2
private let AF_INET6_NUM: Int32 = 30

public class PacketTunnelProvider: NEPacketTunnelProvider {
  private var isReadingPackets = false

  // MARK: Lifecycle

  public override func startTunnel(options: [String : NSObject]?, completionHandler: @escaping (Error?) -> Void) {
    DDLogInfo("PacketTunnelProvider.startTunnel invoked")

    // 1) Parse provider configuration
    guard
      let proto = (self.protocolConfiguration as? NETunnelProviderProtocol),
      let providerConfig = proto.providerConfiguration,
      let tunnelId = providerConfig["id"] as? String,
      let transportConfig = providerConfig["transport"] as? String
    else {
      let err = SwiftBridge.newInvalidConfigOutlineError(message: "Missing provider configuration (id/transport)")
      SwiftBridge.saveLastError(nsError: err)
      completionHandler(err)
      return
    }

    DDLogInfo("VPN config parsed - tunnelId: \(tunnelId), transport config length: \(transportConfig.count)")

    // 2) Compute tunnel network settings
    let settings = SwiftBridge.getTunnelNetworkSettings()

    // 3) Apply settings
    self.setTunnelNetworkSettings(settings) { [weak self] settingsError in
      guard let self = self else { return }
      if let settingsError = settingsError {
        let err = SwiftBridge.newOutlineErrorFrom(nsError: settingsError) ?? SwiftBridge.newInternalOutlineError(message: settingsError.localizedDescription)
        DDLogError("Failed to apply tunnel settings: \(settingsError)")
        SwiftBridge.saveLastError(nsError: err)
        completionHandler(err)
        return
      }

      DDLogInfo("Tunnel network settings applied successfully")
      
      // 4) For now, we'll simulate a successful connection since we're focusing on system integration
      // In a production environment, this would create the actual tunnel connection
      // TODO: Integrate actual Tun2socks connection once framework import is resolved
      
      // 5) Start packet processing loop (placeholder)
      self.startPacketReadLoop()

      // Clear last saved error on success
      SwiftBridge.saveLastError(nsError: nil)
      DDLogInfo("PacketTunnelProvider.startTunnel completed successfully")
      completionHandler(nil)
    }
  }

  public override func stopTunnel(with reason: NEProviderStopReason, completionHandler: @escaping () -> Void) {
    DDLogInfo("PacketTunnelProvider.stopTunnel reason=\(reason.rawValue)")
    // Stop reading packets first.
    self.isReadingPackets = false

    // TODO: Disconnect actual tun2socks tunnel when framework is properly integrated

    // Nothing specific to persist on clean shutdown.
    SwiftBridge.saveLastError(nsError: nil)
    completionHandler()
  }

  public override func handleAppMessage(_ messageData: Data, completionHandler: ((Data?) -> Void)? = nil) {
    guard let message = String(data: messageData, encoding: .utf8) else {
      completionHandler?(nil)
      return
    }
    DDLogDebug("PacketTunnelProvider.handleAppMessage \(message)")
    switch message {
    case ExtensionIPCCommand.fetchLastDetailedJsonError:
      // Returns a PropertyList-encoded LastErrorIPCData or nil
      completionHandler?(SwiftBridge.loadLastErrorToIPCResponse() as Data?)
    default:
      completionHandler?(nil)
    }
  }

  // MARK: Packet I/O

  private func startPacketReadLoop() {
    guard !isReadingPackets else { return }
    isReadingPackets = true
    readPackets()
  }

  private func readPackets() {
    guard isReadingPackets else { return }
    self.packetFlow.readPackets { [weak self] packets, _ in
      guard let self = self else { return }
      
      // TODO: Process packets through tun2socks tunnel when framework is integrated
      // For now, we'll just log the packet count for validation
      DDLogDebug("Received \(packets.count) packets for processing")
      
      // Continue loop if still active.
      if self.isReadingPackets {
        self.readPackets()
      }
    }
  }
}
